internal/types/document.go

```diff
@@ -35,12 +35,18 @@ type document interface {
 
 // Document represents BSON document: an ordered collection of fields
 // (key/value pairs where key is a string and value is any BSON value).
+//
+// Data documents (that are stored in the backend) have a special RecordID property
+// that is not a field and can't be accessed by most methods.
+// It use used to locate the document in the backend.
 type Document struct {
-	fields []field
-	frozen bool
+	fields   []field
+	frozen   bool
+	recordID Timestamp
```

Comment:
What are the benefits of using timestamp for this field? If we want to remove the oldest document then why can't we just use incremental id for such task?

Also, what if more than one documents were inserted in a single second? Which one should be removed first?

Reply:
The uniqueness will need to be ensured by backend, using index possibly. If new insert is not unique, another timestamp can be re-generated and set, hence recordID can be set also on frozen document I think. So neither should be remove.

Incremental ID generation will depend heavily on each backend, or if we don't want to depend on backend, we need to fetch last incremental id, then compute a new one. So transaction + more code maybe involved, so I think using timestamp is easier :hugs:

Reply:
The index will be used, yes.

`types.Timestamp` includes both the time… stamp 🥁 and a counter. Even if the second is the same, the counter value will be different. It is not the same as `time.Time`.

There are cases when we will need to know Record ID before we insert the row to the table.

Reply:
I see, so it's more guarantee of uniqueness than I initially read.